File: leetcode-solutions/0200-number-of-islands/solution.py
class Solution:
    def numIslands(self, grid: List[List[str]]) -> int:
        # without set o(1) space
        # visited = set()
        ROWS, COLS = len(grid), len(grid[0])
        count = 0

        def dfs(r, c):
            if r >= ROWS or r < 0 or c >= COLS or c < 0 or grid[r][c] == "0":
                return
            
            grid[r][c] = "0"
            # visited.add((r,c))
            dfs(r+1, c)
            dfs(r-1, c)
            dfs(r, c+1)
            dfs(r, c-1)

        for r in range(ROWS):
            for c in range(COLS):
                if grid[r][c] == "1":
                    dfs(r,c)
                    count += 1
        return count

File: leetcode-solutions/0200-number-of-islands/solution.java
class Solution {
    
    private Set<Integer> visited;
    private char[][] grid;
    private int rows, cols;

    public int numIslands(char[][] grid) {
        this.grid = grid;
        rows = grid.length;
        cols = grid[0].length;
        visited = new HashSet<>();
        int count = 0;

        for (int r = 0; r < rows; r++) {
            for (int c = 0; c < cols; c++) {
                if (grid[r][c] == '1' && !visited.contains(r * cols + c)) {
                    count++;
                    dfs(r, c);
                }
            }
        }
        return count;
    }

    private void dfs(int r, int c) {
        if (r < 0 || r >= rows || c < 0 || c >= cols || grid[r][c] == '0' || visited.contains(r * cols + c)) {
            return;
        }
        visited.add(r * cols + c);
        dfs(r + 1, c);
        dfs(r - 1, c);
        dfs(r, c + 1);
        dfs(r, c - 1);
    }
}
